src/librbd/ExclusiveLock.cc

```diff
@@ -450,24 +447,29 @@ void ExclusiveLock<I>::handle_acquire_lock(int r) {
     ldout(cct, 5) << "successfully acquired exclusive lock" << dendl;
   }
 
-  {
-    m_lock.Lock();
-    assert(m_state == STATE_ACQUIRING ||
-           m_state == STATE_POST_ACQUIRING ||
-           m_state == STATE_WAITING_FOR_PEER);
-
-    Action action = get_active_action();
-    assert(action == ACTION_TRY_LOCK || action == ACTION_REQUEST_LOCK);
-    if (action == ACTION_REQUEST_LOCK && r < 0 && r != -EBLACKLISTED &&
-        r != -EPERM && r != -EROFS) {
-      m_state = STATE_WAITING_FOR_PEER;
-      m_lock.Unlock();
+  Mutex::Locker locker(m_lock);
+  handle_acquire_lock(r, m_lock);
+}
 
-      // request the lock from a peer
-      m_image_ctx.image_watcher->notify_request_lock();
-      return;
-    }
-    m_lock.Unlock();
+template <typename I>
+void ExclusiveLock<I>::handle_acquire_lock(int r, Mutex &lock) {
+  assert(m_lock.is_locked());
+  assert(m_state == STATE_ACQUIRING ||
+         m_state == STATE_POST_ACQUIRING ||
+         m_state == STATE_WAITING_FOR_PEER);
+
+  Action action = get_active_action();
+  assert(action == ACTION_TRY_LOCK || action == ACTION_REQUEST_LOCK);
+  if (action == ACTION_REQUEST_LOCK && r < 0 && r != -EBLACKLISTED &&
+      r != -EPERM && r != -EROFS) {
+    m_state = STATE_WAITING_FOR_PEER;
+
+    // request the lock from a peer
+    auto ctx = new FunctionContext([this](int r) {
+        m_image_ctx.image_watcher->notify_request_lock();
+      });
+    m_image_ctx.op_work_queue->queue(ctx, 0);
```

Comment:
I would imagine that this would now potentially create race conditions within the unit test (as evidenced by the fact that you had to re-order the call expectations).